Add example of track-to-track fusion

ekhunter123 opened the pull request:
Add example of track-to-track fusion using the new Bayesian formulation of the covariance intersection algorithm from Clark and Campbell [(link)](https://www.techrxiv.org/articles/preprint/Integrating_Covariance_Intersection_into_Bayesian_multi-target_tracking_filters/19351931).

I have run Sphinx to check the formatting, here is what it looks like as a PDF: 
[Track2Track_Fusion_Example.pdf](https://github.com/dstl/Stone-Soup/files/8992963/Track2Track_Fusion_Example.pdf)

This work was done with Daniel Clark, Sean O'Rourke (@orourkesm-afrl), and Bhashyam Balaji (@bhashyambalaji).

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+#     sensors outputs tracks instead of raw measurements.
+#   * Create a GM-PHD tracker that will perform measurement fusion, using all measurements from
+#     both radars. This is created to compare with the covariance intersection method.
+#   * Define the ChernoffUpdater class, which implements the covariance intersection
```
```suggestion
#   * Define the :class:`ChernoffUpdater` class, which implements the covariance intersection
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+#     both radars. This is created to compare with the covariance intersection method.
+#   * Define the ChernoffUpdater class, which implements the covariance intersection
+#   * Define a Detector class which converts tracks into Gaussian measurements
+#   * Create a GM-PHD tracker using the ChernoffUpdater for the covariance intersection
```
```suggestion
#   * Create a GM-PHD tracker using the :class:`ChernoffUpdater` for the covariance intersection
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+
+# %%
+# The first radar will be airborne, at an altitude of approximately 3000 m. It makes detections with
+# an elevation, bearing, and range measurement model. By setting the max_range to 3500, we can
```
```suggestion
# an elevation, bearing, and range measurement model. By setting the `max_range` to 3500, we can
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+
+# %%
+# Now we can pass the platforms into a detection simulator. At each timestep, the simulator will
+# return the detections from the sensor1 platform, then the detections from the sensor2 platform.
```
```suggestion
# return the detections from the `sensor1_platform`, then the detections from the `sensor2_platform`.
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+# %%
+# 5: Define the Chernoff Updater Class
+# ------------------------------------
+# The ChernoffUpdater is where we will implement the covariance intersection. As an
```
```suggestion
# The :class:`ChernoffUpdater` is where we will implement the covariance intersection. As an
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+# 5: Define the Chernoff Updater Class
+# ------------------------------------
+# The ChernoffUpdater is where we will implement the covariance intersection. As an
+# :class:`~.Updater`, the ChernoffUpdater will need functions `predict_measurement()` and
```
```suggestion
# :class:`~.Updater`, the :class:`ChernoffUpdater` will need the functions `predict_measurement()` and
```

nperree-dstl commented on docs/examples/Track2Track_Fusion_Tutorial.py:
```diff
@@ -0,0 +1,876 @@
+# invited to run it yourself and remove some of the items from the plot. Or, if you are running
+# the notebook locally, add the next line to your code to make it an interactive plot.
+#
+# %matplotlib widget
```
Is this needed in the docs?

ekhunter123 replied:
Maybe not? I added it as an excuse for the very messy plot, it is a lot easier to see using the widget mode. But I can remove it if you think it doesn't quite fit!

ekhunter123 replied:
I've changed the wording a little. I think it is still useful to have in an example like this. I'm hoping it will encourage users to explore the plots and results more closely on their own, as that is the best way to fully understand the material.

nperree-dstl replied:
No it's okay, if it's needed for the plotting leave it in - I just wasn't sure if it did anything on the documentation version

nperree-dstl reviewed (commented):
Tutorial looks good! Just some really small documentation comments from looking at the pdf version.

ekhunter123 commented:
Now that the ChernoffUpdater and the track feeder are merged into main, I could remove their class definitions in this example. Or is it helpful to have them here for clarity?

nperree-dstl commented:
> Now that the ChernoffUpdater and the track feeder are merged into main, I could remove their class definitions in this example. Or is it helpful to have them here for clarity?

Yeah I would probably remove them from the example as it is quite a long one and let people explore the classes themselves if they want to

ekhunter123 commented:
I removed the class definitions for the ChernoffUpdater and the Tracks2GaussianDetectionFeeder, and reworked the explanation to accommodate this. Here is the latest Sphinx rendering

[T2TFusionExample.pdf](https://github.com/dstl/Stone-Soup/files/9243102/T2TFusionExample.pdf)

sdhiscocks commented on docs/examples/Track2Track_Fusion_Example.py:
```diff
@@ -0,0 +1,784 @@
+
+plotter1.fig.show()
+
+track = track_fusion_tracks.pop()
```
Building on CircleCI currently fails as `track_fusion_tracks` is an empty set.

ekhunter123 replied:
Ah I see, thanks Steve! I added an if statement to check if the track_fusion_tracks has any tracks in it before plotting. I've run the notebook so many times on my own, and it very rarely happens that there are no tracks. So I think this is an adequate solution.

sdhiscocks commented:
I'm not sure why, but when running on CircleCI, no tracks are produced. But it works fine locally.

I'll rebase on `main` to see if that resolves it.

sdhiscocks commented:
So the issue was in certain configurations, the prune threshold was slightly too high. I recreated this on another Linux system running Python 3.10. I suspect this is caused by floating point precision discrepancy, either with Linux, Python 3.10, or version of NumPy or 3rd party libraries. Increase the threshold seems to have resolved the issue.
